src/meta/src/stream/stream_graph.rs

```diff
@@ -184,359 +175,399 @@ impl StreamActorBuilder {
+        downstream_actors: &[GlobalActorId],
+        downstream_actor_mapping: ActorMapping,
+    ) -> Dispatcher {
+        Dispatcher {
+            r#type: DispatcherType::Hash as _,
+            column_indices: column_indices.to_vec(),
+            hash_mapping: Some(downstream_actor_mapping.to_protobuf()),
+            dispatcher_id: downstream_fragment_id.as_global_id() as u64,
+            downstream_actor_id: GlobalActorIds(downstream_actors.to_vec()).as_global_ids(), // TODO
         }
+    }
 
-        // otherwise, make m * n links between actors.
-
-        assert!(
-            !same_worker_node,
-            "same_worker_node only applies to 1v1 dispatchers."
-        );
-
-        // update actors to have dispatchers, link upstream -> downstream.
-        upstream_actor_ids.iter().for_each(|upstream_id| {
-            self.actor_builders
-                .get_mut(upstream_id)
-                .unwrap()
-                .add_dispatcher(
-                    dispatch_strategy.clone(),
-                    dispatch_id,
-                    OrderedActorLink(downstream_actor_ids.to_vec()),
-                    same_worker_node,
-                );
-        });
+    /// Create a new dispatcher for non-hash types.
+    fn new_normal_dispatcher(
+        dispatcher_type: DispatcherType,
+        downstream_fragment_id: GlobalFragmentId,
+        downstream_actors: &[GlobalActorId],
+    ) -> Dispatcher {
+        assert_ne!(dispatcher_type, DispatcherType::Hash);
+        Dispatcher {
+            r#type: dispatcher_type as _,
+            column_indices: Vec::new(),
+            hash_mapping: None,
+            dispatcher_id: downstream_fragment_id.as_global_id() as u64,
+            downstream_actor_id: GlobalActorIds(downstream_actors.to_vec()).as_global_ids(), // TODO
+        }
+    }
 
-        // update actors to have upstreams, link downstream <- upstream.
-        downstream_actor_ids.iter().for_each(|downstream_id| {
-            self.actor_builders
-                .get_mut(downstream_id)
-                .unwrap()
-                .upstreams
-                .try_insert(
-                    exchange_operator_id,
-                    StreamActorUpstream {
-                        actors: OrderedActorLink(upstream_actor_ids.to_vec()),
-                        fragment_id: upstream_fragment_id,
-                        same_worker_node,
-                    },
-                )
-                .unwrap_or_else(|_| {
-                    panic!(
-                        "duplicated exchange input {} for actors {:?} -> {:?}",
-                        exchange_operator_id, upstream_actor_ids, downstream_actor_ids
-                    )
-                });
-        });
+    /// Add the new dispatcher for an actor.
+    ///
+    /// - If the actor is to be built, the dispatcher will be added to the actor builder.
+    /// - If the actor is an external actor, the dispatcher will be added to the external changes.
```

Comment:
By recording external changes, we can directly get the new dispatchers to add to the original `Materialize`.